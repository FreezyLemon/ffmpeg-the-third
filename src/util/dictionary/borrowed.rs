use std::fmt;
use std::marker::PhantomData;

use super::{impls, Dictionary, Flags, Iter};
use crate::ffi::*;

pub struct DictionaryRef<'d> {
    ptr: *const AVDictionary,
    _marker: PhantomData<&'d AVDictionary>,
}

impl<'d> DictionaryRef<'d> {
    pub unsafe fn from_raw(ptr: *const AVDictionary) -> Self {
        DictionaryRef {
            ptr,
            _marker: PhantomData,
        }
    }

    pub fn as_ptr(&self) -> *const AVDictionary {
        self.ptr
    }
}

impl<'d> DictionaryRef<'d> {
    pub fn get(&self, key: &str) -> Option<&'d str> {
        unsafe { impls::get(self.as_ptr(), key, Flags::empty()) }
    }

    pub fn iter(&self) -> Iter {
        Iter::new(self.as_ptr())
    }

    pub fn to_owned(&self) -> Dictionary {
        self.iter().collect()
    }
}

impl<'a> IntoIterator for &'a DictionaryRef<'a> {
    type Item = (&'a str, &'a str);
    type IntoIter = Iter<'a>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

impl<'a> fmt::Debug for DictionaryRef<'a> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_map().entries(self.iter()).finish()
    }
}
