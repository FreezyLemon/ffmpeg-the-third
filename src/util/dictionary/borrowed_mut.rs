use core::fmt;
use std::marker::PhantomData;

use crate::ffi::*;

pub struct DictionaryMut<'d> {
    ptr: *mut AVDictionary,
    _marker: PhantomData<&'d mut AVDictionary>,
}

impl<'d> DictionaryMut<'d> {
    pub unsafe fn from_raw(ptr: *mut AVDictionary) -> Self {
        DictionaryMut {
            ptr,
            _marker: PhantomData,
        }
    }

    pub fn as_ptr(&self) -> *const AVDictionary {
        self.ptr
    }

    pub fn as_mut_ptr(&mut self) -> *mut AVDictionary {
        self.ptr
    }

    pub fn as_ref(&self) -> super::DictionaryRef<'d> {
        unsafe { super::DictionaryRef::from_raw(self.as_ptr()) }
    }
}

impl<'d> fmt::Debug for DictionaryMut<'d> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.as_ref().fmt(f)
    }
}
