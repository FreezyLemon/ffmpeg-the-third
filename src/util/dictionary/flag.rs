use crate::ffi::*;
use libc::c_int;

bitflags::bitflags! {
    pub struct Flags: c_int {
        /// Only get an entry with exact-case key match.
        const MATCH_CASE        = AV_DICT_MATCH_CASE;

        /// Return first entry in a dictionary whose first part corresponds to the
        /// search key, ignoring the suffix of the found key string.
        const IGNORE_SUFFIX     = AV_DICT_IGNORE_SUFFIX;

        /// Take ownership of a key that's been allocated with av_malloc() or
        /// another memory allocation function.
        const DONT_STRDUP_KEY   = AV_DICT_DONT_STRDUP_KEY;

        /// Take ownership of a value that's been allocated with av_malloc()
        /// or another memory allocation function.
        const DONT_STRDUP_VAL   = AV_DICT_DONT_STRDUP_VAL;

        /// Don't overwrite existing entries.
        const DONT_OVERWRITE    = AV_DICT_DONT_OVERWRITE;

        /// If the entry already exists, append to it.
        ///
        /// Note that no delimiter is added, the strings are simply concatenated.
        const APPEND            = AV_DICT_APPEND;

        /// Allow to store several equal keys in the dictionary.
        const MULTIKEY          = AV_DICT_MULTIKEY;
    }
}
