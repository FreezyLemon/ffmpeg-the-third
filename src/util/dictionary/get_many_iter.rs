use std::ffi::CString;

use libc::c_int;

use super::{Dictionary, DictionaryEntry};

pub struct DictionaryGetManyIter<'d: 'e, 'e>
{
    dict: Dictionary<'d>,
    key: CString,
    flags: c_int,
    prev: Option<DictionaryEntry<'e>>,
}

impl<'d: 'e, 'e> DictionaryGetManyIter<'d, 'e> {
    pub fn new(dict: Dictionary<'d>, key: CString, flags: c_int) -> Self {
        Self {
            dict,
            key,
            flags,
            prev: None,
        }
    }
}

impl<'d: 'e, 'e> Iterator for DictionaryGetManyIter<'d, 'e> {
    type Item = DictionaryEntry<'e>;

    fn next(&mut self) -> Option<Self::Item>
    {
        match self.dict.get(&self.key, self.prev, self.flags) {
            Ok(entry) => {
                self.prev = Some(entry);
                Some(entry)
            }
            Err(_) => None,
        }
    }
}
