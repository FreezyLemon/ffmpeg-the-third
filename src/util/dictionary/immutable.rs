use std::fmt;
use std::marker::PhantomData;

use super::Iter;
use crate::ffi::*;

pub struct DictionaryRef<'d> {
    ptr: *const AVDictionary,
    _marker: PhantomData<&'d AVDictionary>,
}

impl<'d> DictionaryRef<'d> {
    pub unsafe fn from_raw(ptr: *const AVDictionary) -> Self {
        DictionaryRef {
            ptr,
            _marker: PhantomData,
        }
    }

    pub fn as_ptr(&self) -> *const AVDictionary {
        self.ptr
    }
}

impl<'a> IntoIterator for &'a DictionaryRef<'a> {
    type Item = (&'a str, &'a str);
    type IntoIter = Iter<'a>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

impl<'a> fmt::Debug for DictionaryRef<'a> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_map().entries(self.iter()).finish()
    }
}
