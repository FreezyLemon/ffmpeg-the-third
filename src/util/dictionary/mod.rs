mod flag;
pub use flag::Flags;

mod immutable;
pub use immutable::DictionaryRef;

mod impls;

mod mutable;
pub use mutable::DictionaryMut;

mod owned;
pub use owned::Dictionary;

mod iter;
pub use iter::Iter;

#[cfg(test)]
mod tests;

#[macro_export]
macro_rules! dict {
    ($($key:expr => $value:expr),* $(,)*) => ({
        let mut dict = ::ffmpeg::Dictionary::new();

        $(
            dict.set($key, $value);
        )*

        dict
    });
}
