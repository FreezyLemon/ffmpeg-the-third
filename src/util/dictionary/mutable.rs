use std::fmt;

use crate::ffi::*;

pub struct DictionaryMut<'d> {
    ptr: &'d mut *mut AVDictionary,
}

impl<'d> DictionaryMut<'d> {
    pub unsafe fn from_raw(ptr: &'d mut *mut AVDictionary) -> Self {
        DictionaryMut { ptr }
    }

    pub fn as_ptr(&self) -> *const AVDictionary {
        *self.ptr
    }

    pub fn as_mut_ptr(&mut self) -> &mut *mut AVDictionary {
        self.ptr
    }

    pub fn as_ref(&self) -> super::DictionaryRef<'d> {
        unsafe { super::DictionaryRef::from_raw(self.as_ptr()) }
    }
}

impl<'d> fmt::Debug for DictionaryMut<'d> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.as_ref().fmt(f)
    }
}
