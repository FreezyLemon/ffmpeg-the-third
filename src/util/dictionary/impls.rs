use std::ffi::CString;

use super::flag::Flags;
use crate::{ffi::*, utils};

pub fn set(dict: &mut *mut AVDictionary, key: &str, value: &str, flags: Flags) {
    let key = CString::new(key).unwrap();
    let value = CString::new(value).unwrap();

    unsafe {
        if av_dict_set(dict, key.as_ptr(), value.as_ptr(), flags.bits()) < 0 {
            panic!("out of memory");
        }
    }
}

pub unsafe fn get<'d>(dict: *const AVDictionary, key: &str, flags: Flags) -> Option<&'d str> {
    let key = CString::new(key).unwrap();
    unsafe {
        let entry = av_dict_get(dict, key.as_ptr(), std::ptr::null_mut(), flags.bits());

        if entry.is_null() {
            None
        } else {
            Some(utils::str_from_c_ptr((*entry).value))
        }
    }
}
