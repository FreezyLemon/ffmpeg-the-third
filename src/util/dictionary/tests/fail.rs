use ffmpeg_the_third::dictionary::*;

fn main() {}

pub fn drop_owned_while_ref_alive() {
    let dict = Dictionary::new();
    let dict_ref = dict.as_ref();
    drop(dict);
    println!("len: {}", dict_ref.len());
}

pub fn drop_owned_while_mut_alive() {
    let mut dict = Dictionary::new();
    let dict_mut = dict.as_mut();
    drop(dict);
    println!("len: {}", dict_mut.len());
}
