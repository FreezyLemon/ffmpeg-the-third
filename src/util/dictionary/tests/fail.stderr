error[E0505]: cannot move out of `dict` because it is borrowed
 --> src/util/dictionary/tests/fail.rs:8:10
  |
6 |     let dict = Dictionary::new();
  |         ---- binding `dict` declared here
7 |     let dict_ref = dict.as_ref();
  |                    ---- borrow of `dict` occurs here
8 |     drop(dict);
  |          ^^^^ move out of `dict` occurs here
9 |     println!("len: {}", dict_ref.len());
  |                         -------- borrow later used here
  |
help: consider cloning the value if the performance cost is acceptable
  |
7 |     let dict_ref = dict.clone().as_ref();
  |                        ++++++++

error[E0505]: cannot move out of `dict` because it is borrowed
  --> src/util/dictionary/tests/fail.rs:15:10
   |
13 |     let mut dict = Dictionary::new();
   |         -------- binding `dict` declared here
14 |     let dict_mut = dict.as_mut();
   |                    ---- borrow of `dict` occurs here
15 |     drop(dict);
   |          ^^^^ move out of `dict` occurs here
16 |     println!("len: {}", dict_mut.len());
   |                         -------- borrow later used here
